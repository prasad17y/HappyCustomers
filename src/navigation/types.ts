import {NavigationProp} from '@react-navigation/native';

export type RootStackParamList = {
  UserList: undefined;
  AddUser: undefined;
};

export type RootStackNavigationProp = NavigationProp<RootStackParamList>;
