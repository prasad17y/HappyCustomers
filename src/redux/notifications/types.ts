export type ToastType = 'success' | 'error' | 'info';

export interface ToastPayload {
  message: string;
  type: ToastType;
}

export interface NotificationsState {
  toast: ToastPayload | null;
}

export const SHOW_TOAST = 'SHOW_TOAST';
export const CLEAR_TOAST = 'CLEAR_TOAST';

interface ShowToastAction {
  type: typeof SHOW_TOAST;
  payload: ToastPayload;
}

interface ClearToastAction {
  type: typeof CLEAR_TOAST;
}

export type NotificationsActionTypes = ShowToastAction | ClearToastAction;
