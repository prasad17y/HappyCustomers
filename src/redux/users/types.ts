export interface UsersState {
  isSyncing: boolean;
  syncError: string | null;
  lastSyncTimestamp: string | null;
}

export const SYNC_USERS_REQUEST = 'SYNC_USERS_REQUEST';
export const SYNC_USERS_SUCCESS = 'SYNC_USERS_SUCCESS';
export const SYNC_USERS_FAILURE = 'SYNC_USERS_FAILURE';
export const CLEAR_SYNC_ERROR = 'CLEAR_SYNC_ERROR';

interface SyncUsersRequestAction {
  type: typeof SYNC_USERS_REQUEST;
  payload?: {forceRefresh: boolean};
}

interface SyncUsersSuccessAction {
  type: typeof SYNC_USERS_SUCCESS;
  payload: {didFetch: boolean};
}

interface SyncUsersFailureAction {
  type: typeof SYNC_USERS_FAILURE;
  payload: string;
}

interface ClearSyncErrorAction {
  type: typeof CLEAR_SYNC_ERROR;
}

export type UsersActionTypes =
  | SyncUsersRequestAction
  | SyncUsersSuccessAction
  | SyncUsersFailureAction
  | ClearSyncErrorAction;
