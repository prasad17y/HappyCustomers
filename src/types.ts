export interface UserType {
  id: string;
  name: string;
  email: string | null;
  role: string;
}
