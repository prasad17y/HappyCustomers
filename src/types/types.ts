export interface UserType {
  id: string;
  name: string;
  email: string | null;
  role: string;
}

export enum Role {
  ADMIN = 'ADMIN',
  MANAGER = 'MANAGER',
}

export interface Tab {
  key: string;
  title: string;
}

export interface FilterTab extends Tab {
  role?: Role;
}
