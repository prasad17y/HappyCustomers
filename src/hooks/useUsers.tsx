import {useState, useEffect} from 'react';
import {Q} from '@nozbe/watermelondb';
import {database} from '../db';
import UserModel from '../db/models/UserModel';
import {UserType, Role} from '../types/types';

// queries the database, and subscribes to changes
export const useUsers = (roleFilter?: Role, searchQuery?: string) => {
  const [users, setUsers] = useState<UserType[]>([]);

  useEffect(() => {
    const usersCollection = database.collections.get<UserModel>('users');
    const queryConditions: any[] = [Q.sortBy('name', Q.asc)];

    // Add role filter
    if (roleFilter) {
      queryConditions.push(Q.where('role', roleFilter));
    }

    // Add search filter
    if (searchQuery && searchQuery.trim().length > 0) {
      const sanitizedQuery = searchQuery.trim();
      // Case-insensitive search for names that start with the query
      queryConditions.push(
        Q.where('name', Q.like(`${Q.sanitizeLikeString(sanitizedQuery)}%`)),
      );
    }

    const query = usersCollection.query(...queryConditions);

    const subscription = query.observe().subscribe(latestUserModels => {
      const userTypes: UserType[] = latestUserModels.map(user => ({
        id: user.id,
        name: user.name,
        email: user.email,
        role: user.role,
      }));
      setUsers(userTypes);
    });

    return () => subscription.unsubscribe();
  }, [roleFilter, searchQuery]); // Rerun if filter or search changes

  return users;
};
