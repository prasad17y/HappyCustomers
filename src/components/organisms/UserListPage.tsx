import React, {useMemo, useCallback} from 'react';
import {View, ActivityIndicator, StyleSheet, Text} from 'react-native';
import {Role, UserType} from '../../types/types';
import {useUsers} from '../../hooks/useUsers';
import List from '../molecules/List';
import UserListItem from '../molecules/UserListItem';
import {useIsFocused} from '@react-navigation/native';

interface UserListPageProps {
  role?: Role;
  searchQuery: string;
  isRefreshing: boolean;
  onRefresh: () => void;
  isSelected: boolean;
}

interface Section {
  title: string;
  data: UserType[];
}

const UserListPage: React.FC<UserListPageProps> = ({
  role,
  searchQuery,
  isRefreshing,
  onRefresh,
  isSelected,
}) => {
  const isScreenFocused = useIsFocused();

  const {users, error} = useUsers(
    role,
    searchQuery,
    isScreenFocused && isSelected, // run new query only if this is true
  );

  const groupedUsers = useMemo(() => {
    if (users === undefined) {
      return [];
    }
    return users.reduce<Section[]>((acc, user) => {
      const initial = user.name.charAt(0).toUpperCase();
      const lastSection = acc.length > 0 ? acc[acc.length - 1] : undefined;

      if (lastSection && lastSection.title === initial) {
        lastSection.data.push(user);
      } else {
        acc.push({title: initial, data: [user]});
      }
      return acc;
    }, []);
  }, [users]);

  const keyExtractor = useCallback((item: UserType) => item.id, []);

  const {
    renderItem,
    renderEmptyComponent,
    renderSectionHeader,
    renderFooterComponent,
  } = useMemo(() => {
    return {
      renderItem: ({item}: {item: UserType}) => <UserListItem user={item} />,
      renderEmptyComponent: () => (
        <View style={styles.emptyContainer}>
          <Text style={styles.emptyText}>No results found</Text>
        </View>
      ),
      renderSectionHeader: ({section}: {section: Section}) => (
        <View style={styles.sectionHeaderContainer}>
          <Text style={styles.sectionHeaderText}>{section.title}</Text>
        </View>
      ),
      renderFooterComponent: () => <View style={styles.footer} />,
    };
  }, []);

  if (error) {
    return (
      <View style={styles.centered}>
        <Text style={styles.errorText}>
          An error occurred while loading users.
        </Text>
      </View>
    );
  }

  if (users === undefined) {
    return (
      <View style={styles.centered}>
        <ActivityIndicator />
      </View>
    );
  }

  return (
    <List
      sections={groupedUsers}
      renderItem={renderItem}
      renderSectionHeader={renderSectionHeader}
      refreshing={isRefreshing}
      onRefresh={onRefresh}
      keyExtractor={keyExtractor}
      ListEmptyComponent={renderEmptyComponent}
      ListFooterComponent={renderFooterComponent}
    />
  );
};

const styles = StyleSheet.create({
  centered: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  sectionHeaderContainer: {
    backgroundColor: '#f7f7f7',
    paddingVertical: 8,
    paddingHorizontal: 16,
  },
  sectionHeaderText: {
    fontSize: 14,
    fontWeight: 'bold',
    color: '#3c4040',
  },
  emptyContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  emptyText: {
    fontSize: 16,
    color: '#888',
  },
  errorText: {
    fontSize: 16,
    color: 'red',
    textAlign: 'center',
    padding: 20,
  },
  footer: {
    height: 100,
  },
});

export default React.memo(UserListPage);
