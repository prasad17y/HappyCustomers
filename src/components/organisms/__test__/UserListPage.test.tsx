import React from 'react';
import {render, screen} from '@testing-library/react-native';
import UserListPage from '../UserListPage';
import * as useUsersHook from '../../../hooks/useUsers';
import {Role} from '../../../types/types';
import {useIsFocused} from '@react-navigation/native';

// Mock the useIsFocused hook from react-navigation
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useIsFocused: jest.fn(),
}));

// Mock the entire useUsers hook
jest.mock('../../../hooks/useUsers');
const useUsersMock = useUsersHook.useUsers as jest.Mock;

// Mock the List molecule to simplify the test and check the props it receives
jest.mock('../../molecules/List', () => {
  const React = require('react');
  const {View, Text} = require('react-native');
  return (props: {sections: any[]; refreshing: boolean}) => (
    <View testID="mock-list">
      <Text>{`Sections: ${props.sections.length}`}</Text>
      <Text>{`Refreshing: ${props.refreshing}`}</Text>
    </View>
  );
});

describe('UserListPage', () => {
  const defaultProps = {
    searchQuery: '',
    isRefreshing: false,
    onRefresh: jest.fn(),
    isSelected: true,
  };

  beforeEach(() => {
    // Reset mocks before each test
    useUsersMock.mockClear();
    // Provide a default mock return value for useIsFocused for all tests
    (useIsFocused as jest.Mock).mockReturnValue(true);
  });

  it('should render the ActivityIndicator while the user list is loading', () => {
    // Arrange: Mock the hook to return the "loading" state (undefined)
    useUsersMock.mockReturnValue({users: undefined, error: null});

    render(<UserListPage {...defaultProps} />);

    // Assert: Check that the ActivityIndicator is visible
    expect(screen.getByTestId('activity-indicator')).toBeTruthy();
    expect(screen.queryByTestId('mock-list')).toBeNull();
  });

  it('should render an error message if the hook returns an error', () => {
    // Arrange: Mock the hook to return an error
    useUsersMock.mockReturnValue({
      users: undefined,
      error: new Error('DB Error'),
    });

    render(<UserListPage {...defaultProps} />);

    // Assert: Check that the error message is displayed
    expect(
      screen.getByText('An error occurred while loading users.'),
    ).toBeTruthy();
  });

  it('should render the list and correctly group users on success', () => {
    // Arrange: Mock the hook to return a list of users
    const mockUsers = [
      {id: '1', name: 'Alice', role: Role.ADMIN},
      {id: '2', name: 'Adam', role: Role.MANAGER},
      {id: '3', name: 'Bob', role: Role.ADMIN},
    ];
    useUsersMock.mockReturnValue({users: mockUsers, error: null});

    render(<UserListPage {...defaultProps} />);

    // Assert: Check that the mock list is rendered and received the correct number of sections
    const list = screen.getByTestId('mock-list');
    expect(list).toBeTruthy();
    // The component should group "Alice" and "Adam" into one section ('A') and "Bob" into another ('B')
    expect(screen.getByText('Sections: 2')).toBeTruthy();
  });

  it('should pass the correct refreshing prop to the List component', () => {
    // Arrange: Mock a successful data return
    useUsersMock.mockReturnValue({users: [], error: null});

    // Render with refreshing set to true
    render(<UserListPage {...defaultProps} isRefreshing={true} />);

    // Assert: Check that the refreshing state is correctly passed down
    expect(screen.getByText('Refreshing: true')).toBeTruthy();
  });

  it('should match the snapshot for a successful render', () => {
    // Arrange: Mock a successful data return
    const mockUsers = [{id: '1', name: 'Charlie', role: Role.MANAGER}];
    useUsersMock.mockReturnValue({users: mockUsers, error: null});

    const tree = render(<UserListPage {...defaultProps} />);

    // Assert: The rendered structure should match the saved snapshot
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
