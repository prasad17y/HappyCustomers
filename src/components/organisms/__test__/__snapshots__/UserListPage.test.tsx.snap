// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserListPage should match the snapshot for a successful render 1`] = `
<View
  testID="mock-list"
>
  <Text>
    Sections: 1
  </Text>
  <Text>
    Refreshing: false
  </Text>
</View>
`;
